### PR TITLE
Application invitations

### PR DESCRIPTION
**Pull request by dandds**
PT story: https://www.pivotaltracker.com/story/show/165061837

This story is technical blocked by [the edit version of the team table](https://www.pivotaltracker.com/story/show/165059841), but we can just drop in the "add new member" link where we need it when that story is done. For now, I've put it in the footer of the view-only version of the table.

Users can now add new application members.

Screen Shot 2019-04-29 at 6 54 01 AM

Screen Shot 2019-04-29 at 6 55 39 AM

Screen Shot 2019-04-29 at 6 56 01 AM

### CODE REVIEW

File: atst/domain/invitations.py [hunk 1]
```diff
@@ -103,23 +105,29 @@ def _update_status(cls, invite, new_status):
...
+        invite = cls._get(token)
+        return cls._update_status(invite, InvitationStatus.REVOKED)
 
     @classmethod
     def lookup_by_portfolio_and_user(cls, portfolio, user):
-        portfolio_role = PortfolioRoles.get(portfolio.id, user.id)
+        role = PortfolioRoles.get(portfolio.id, user.id)
 
-        if portfolio_role.latest_invitation is None:
+        if role.latest_invitation is None:
```
Comment (drummond-dds) [inline atst/domain/invitations.py]: Maybe we can raise `NotFoundError` with the invitation class type - `PortfolioInvitation` or `ApplicationInvitation`

File: atst/routes/applications/team.py [hunk 2]
```diff
@@ -42,8 +46,57 @@ def team(application_id):
...
+
+            flash("new_portfolio_member", new_member=member)
+
+        except AlreadyExistsError:
+            return render_template(
+                "error.html", message="There was an error processing your request."
+            )
+    else:
+        pass
+        # TODO: flash error message
```
Comment (drummond-dds) [inline atst/routes/applications/team.py]: Is that in scope for this ticket?

File: atst/services/invitation.py [hunk 3]
```diff
@@ -47,28 +48,46 @@ def update_officer_invitations(user, task_order):
...
-    ):
+    def __init__(self, inviter, member, email, subject="", email_template=None):
         self.inviter = inviter
         self.member = member
         self.email = email
         self.subject = subject
         self.email_template = email_template
 
+        if isinstance(member, PortfolioRole):
+            self.email_template = (
```
Comment (drummond-dds) [inline atst/services/invitation.py]: I wonder if lines 59-63/71-75 and 64-68/76-80 can be DRYer

File: js/lib/input_validations.js [hunk 4]
```diff
@@ -35,7 +35,7 @@ export default {
   },
   email: {
     mask: emailMask,
-    match: /(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])/i,
+    match: /^.+@[^.].*\.[a-z]{2,10}$/,
```
Comment (drummond-dds) [inline js/lib/input_validations.js]: Was this change just to make things simpler? Are there any emails that were previously valid that are no longer?

File: styles/components/_forms.scss [hunk 5]
```diff
@@ -171,3 +179,15 @@
   }
 }
 
+.input__inline-fields  {
+  margin: 1rem 0 1rem 0;
```
Comment (drummond-dds) [inline styles/components/_forms.scss]: Is this the same as `margin: 1rem 0;`

File: styles/components/_forms.scss [hunk 6]
```diff
@@ -171,3 +179,15 @@
...
+
+  &.input__inline-fields--indented {
+    margin-left: 4*$gap;
```
Comment (drummond-dds) [inline styles/components/_forms.scss]: For consistency I think we should do `$gap*n` throughout

File: styles/sections/_application_edit.scss [hunk 7]
```diff
@@ -17,3 +17,34 @@
...
+
+.environment-roles-new {
+  margin-top: 5*$gap;
+  margin-bottom: 8*$gap;
+
+  .usa-input {
+    margin: 2rem 0 2rem 0;
+
+    .usa-input__title-inline {
+      line-height: $hit-area;
```
Comment (drummond-dds) [inline styles/sections/_application_edit.scss]: This made me chuckle

File: templates/fragments/applications/add_new_application_member.html [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    </div>
+  </div>
+  <div class='action-group'>
+    <input
+        type='button'
+        v-on:click="next()"
+        v-bind:disabled="invalid"
+        class='action-group__action usa-button'
+        value='Next'>
+    <a class='action-group__action icon-link icon-link--default' v-on:click="closeModal('{{ new_port_mem }}')">Cancel</a>
```
Comment (drummond-dds) [inline templates/fragments/applications/add_new_application_member.html]: Can use `{{ "common.cancel" | translate }}`

File: js/lib/input_validations.js (hunk 4 above)
Comment (dandds) [reply js/lib/input_validations.js]: Long commit message about it. Basically, the frontend and backend validation patterns should match and they didn't. (You could enter an email address that passed frontend validation but failed backend.)

https://github.com/dod-ccpo/atst/pull/789/commits/ef9b42f559a6b8e96d211b1fb6e683df3e11f8ed

File: atst/domain/invitations.py (hunk 1 above)
Comment (dandds) [reply atst/domain/invitations.py]: Good call. Fixed in cf3225b7447c94b4c8224294d06ac81b8028de34

File: atst/routes/applications/team.py (hunk 2 above)
Comment (dandds) [reply atst/routes/applications/team.py]: There's a similar TODO here: https://github.com/dod-ccpo/atst/blob/master/atst/routes/applications/settings.py#L156

I think we need to decide on and apply a global pattern for these types of cases. If a user is playing by the rules--i.e., using the regular UI with JS enabled in the browser and not sending us random data via cURL--they shouldn't get form validation errors. It's not entirely clear to me what to do in those cases. I made an icebox chore for it: https://www.pivotaltracker.com/story/show/165729650

File: translations.yaml [hunk 9]
```diff
@@ -433,6 +437,14 @@ portfolios:
...
     team_text: Team
     update_button_text: Save
+    members:
+      new:
+        assign_roles: Assign Member Environments and Roles
+        learn_more: Learn more about these roles
+        manage_perms: 'Manage permissions for {application_name}'
+        manage_envs: 'Allow member to <strong>add</strong> and <strong>rename environments</strong> within the application.'
+        delete_envs: 'Allow member to <strong>delete environments</strong> within the application.'
+        manage_team: 'Allow member to <strong>add, update</strong> and <strong>remove members</strong> from the application team.'
```
Comment (drummond-dds) [inline translations.yaml]: More a question for Jill, but whats our position on using the Oxford comma?

File: atst/services/invitation.py (hunk 3 above)
Comment (dandds) [reply atst/services/invitation.py]: Morning brain. 25c05aff56416d5a9f8e630199668122c3a532b6

File: styles/components/_forms.scss (hunk 5 above)
Comment (dandds) [reply styles/components/_forms.scss]: :+1: 6cfd122fe73e6c4917ae4ed5db6b0274ba5991ba

File: atst/forms/data.py [hunk 10]
```diff
@@ -218,3 +218,6 @@
 ]
 
 ENV_ROLES = [(role.value, role.value) for role in CSPRole] + [(None, "No access")]
+FORMATTED_ENV_ROLES = [
+    (role.value, "- {} -".format(role.value)) for role in CSPRole
```
Comment (drummond-dds) [inline atst/forms/data.py]: I'm not sure we are using the `- Something -` pattern with minus signs either side for other forms in the app? Is this consistent?

File: styles/components/_forms.scss (hunk 6 above)
Comment (dandds) [reply styles/components/_forms.scss]: :+1: b4ca2b4c71c6aae297a19c5cb6df103c5d220d6c

File: styles/sections/_application_edit.scss (hunk 7 above)
Comment (dandds) [reply styles/sections/_application_edit.scss]: I _think_ that's the minimum acceptable height someone defined for selectable elements on mobile. It seems to be used to determine that height of a bunch of things in USWDS.

File: styles/components/_forms.scss (hunk 5 above)
Comment (drummond-dds) [reply styles/components/_forms.scss]: I think there might be a few other similar instances we can also hit in this PR

File: templates/fragments/applications/add_new_application_member.html (hunk 8 above)
Comment (drummond-dds) [reply templates/fragments/applications/add_new_application_member.html]: Might be a couple of these in this PR

File: atst/routes/applications/team.py (hunk 2 above)
Comment (drummond-dds) [reply atst/routes/applications/team.py]: Sounds good 👍

File: templates/fragments/applications/add_new_application_member.html (hunk 8 above)
Comment (dandds) [reply templates/fragments/applications/add_new_application_member.html]: :+1: 22bbdd99e4fa862db971f60c303e546854fd5ed4

File: translations.yaml (hunk 9 above)
Comment (dandds) [reply translations.yaml]: @rachel-dtr says yes. d6357f08ea9ec352d2af54d5a605e064dd5d1c04

File: atst/domain/invitations.py [hunk 11]
```diff
@@ -68,29 +70,29 @@ def create(cls, inviter, portfolio_role, email):
...
             raise ExpiredError(invite)
 
         elif invite.is_accepted or invite.is_revoked or invite.is_rejected:
             raise InvitationError(invite)
 
         elif invite.is_pending:  # pragma: no branch
-            Invitations._update_status(invite, InvitationStatus.ACCEPTED)
-            PortfolioRoles.enable(invite.portfolio_role)
+            cls._update_status(invite, InvitationStatus.ACCEPTED)
+            PortfolioRoles.enable(invite.role)
```
Comment (drummond-dds) [inline atst/domain/invitations.py]: In `BaseInvitations` could this also be an `ApplicationRole`?

File: atst/domain/invitations.py [hunk 12]
```diff
@@ -103,23 +105,29 @@ def _update_status(cls, invite, new_status):
 
     @classmethod
     def revoke(cls, token):
-        invite = Invitations._get(token)
-        return Invitations._update_status(invite, InvitationStatus.REVOKED)
+        invite = cls._get(token)
+        return cls._update_status(invite, InvitationStatus.REVOKED)
 
     @classmethod
     def lookup_by_portfolio_and_user(cls, portfolio, user):
```
Comment (drummond-dds) [inline atst/domain/invitations.py]: I wonder if this class method really belongs in `PortfolioInvitations`

File: atst/forms/data.py (hunk 10 above)
Comment (dandds) [reply atst/forms/data.py]: per @rachel-dtr , normalize it. 9653782c47fefda3dc86ac7ac415008a6598a084

File: atst/domain/invitations.py (hunk 11 above)
Comment (dandds) [reply atst/domain/invitations.py]: Yup, fixed in 34d7b1f52c921b5b3abe14b2a45287c2dabad1cf

File: atst/domain/invitations.py (hunk 12 above)
Comment (dandds) [reply atst/domain/invitations.py]: Renamed it to be more general in 34d7b1f52c921b5b3abe14b2a45287c2dabad1cf

File: atst/routes/applications/team.py [hunk 13]
```diff
@@ -1,12 +1,15 @@
...
 
 
 from . import applications_bp
 from atst.domain.environments import Environments
 from atst.domain.applications import Applications
 from atst.domain.authz.decorator import user_can_access_decorator as user_can
-from atst.models.permissions import Permissions
 from atst.domain.permission_sets import PermissionSets
+from atst.forms.application_member import NewForm as NewMemberForm
+from atst.models.permissions import Permissions
```
Comment (montana-mil) [inline atst/routes/applications/team.py]: we can do that `from atst.models import Permissions` trick here!

Comment (dandds) [reply atst/routes/applications/team.py]: :+1: fixed on the amended 4f304d747e3993d105a3857416c14c278f5b858b
